lib/auth/server_monitor.go

```diff
@@ -0,0 +1,169 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+	notificationsv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/notifications/v1"
+	"github.com/gravitational/teleport/api/types"
+)
+
+const (
+	// timeCheckCycle is the period when local times are compared
+	// for collected resources from heartbeats, and notification must be added.
+	timeCheckCycle = 10 * time.Minute
+	// timeShiftThreshold is the duration threshold for triggering a warning
+	// if the time difference exceeds this threshold.
+	timeShiftThreshold = time.Minute
+)
+
+// resourceWithLocalTime is minimal interface for resource with local time.
+type resourceWithLocalTime interface {
+	types.Resource
+
+	// GetLocalTime gets the local time of the current server.
+	GetLocalTime() time.Time
+}
+
+// resourceMonitor stores info about resource and time difference with local time.
+type resourceMonitor struct {
+	resource resourceWithLocalTime
+	diff     time.Duration
+}
+
+// MonitorNodeInfos consumes heartbeat events of other services to periodically
+// compare the auth server time with the time of remote services,
+// and notifying about the time difference between servers.
+func (a *Server) MonitorNodeInfos(ctx context.Context) error {
+	resources := make(map[string]resourceMonitor)
+	ticker := a.clock.NewTicker(timeCheckCycle)
+
+	watcher, err := a.NewWatcher(ctx, types.Watch{
+		Kinds: []types.WatchKind{
+			{Kind: types.KindNode},
+			{Kind: types.KindWindowsDesktopService},
+			{Kind: types.KindKubeServer},
+			{Kind: types.KindAppServer},
+		},
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.Chan():
+			err := upsertGlobalNotification(ctx, a.Services, generateNotificationMessage(resources))
+			if err != nil {
+				slog.ErrorContext(ctx, "can't set notification about time difference", err)
+			}
+		case event := <-watcher.Events():
+			res, ok := event.Resource.(resourceWithLocalTime)
+			if !ok {
+				continue
+			}
+			// Previous version of the servers don't have this parameter, we have to skip them.
+			if res.GetLocalTime().IsZero() {
+				continue
+			}
+
+			if event.Type == types.OpPut {
+				now := a.clock.Now().UnixNano()
+				localTime := res.GetLocalTime().UnixNano()
+				duration := time.Duration(localTime - now)
+
+				if (duration > 0 && duration > timeShiftThreshold) || (duration < 0 && -duration > timeShiftThreshold) {
+					resources[res.GetName()] = resourceMonitor{resource: res, diff: duration}
```

Comment:
yes, we discussed about this with Roman, initially since original issue related to certificates I've set auth service as point of truth. I will change this one to sorting times and informing the difference in general between services

Reply:
Yeah I think we need to be smarter about how we create notifications - maybe we could just create a single notification saying something like "auth server X has time drift of Y compared to nodes node-1, node-2, node-3 and N others", something like that?

We definitely don't want to bombard the user with notifications for each server.